integration-tests/src/test/java/org/apache/druid/tests/indexer/AbstractStreamIndexingTest.java

```diff
@@ -230,6 +230,78 @@ protected void doTestIndexDataStableState(
     }
   }
 
+  protected void doTestIndexDataHandoffEarly(
+      @Nullable Boolean transactionEnabled
+  ) throws Exception
+  {
+    final GeneratedTestConfig generatedTestConfig = new GeneratedTestConfig(
+        INPUT_FORMAT,
+        getResourceAsString(JSON_INPUT_FORMAT_PATH)
+    );
+    try (
+        final Closeable closer = createResourceCloser(generatedTestConfig);
+        final StreamEventWriter streamEventWriter = createStreamEventWriter(config, transactionEnabled)
+    ) {
+      final String taskSpec = generatedTestConfig.getStreamIngestionPropsTransform()
+          .apply(getResourceAsString(SUPERVISOR_WITH_IDLE_BEHAVIOUR_ENABLED_SPEC_TEMPLATE_PATH));
+      LOG.info("supervisorSpec: [%s]\n", taskSpec);
+      // Start supervisor
+      generatedTestConfig.setSupervisorId(indexer.submitSupervisor(taskSpec));
+      LOG.info("Submitted supervisor");
+
+      // Start generating half of the data
+      int secondsToGenerateRemaining = TOTAL_NUMBER_OF_SECOND;
+      int secondsToGenerateFirstRound = TOTAL_NUMBER_OF_SECOND / 2;
+      secondsToGenerateRemaining = secondsToGenerateRemaining - secondsToGenerateFirstRound;
+      final StreamGenerator streamGenerator = new WikipediaStreamEventStreamGenerator(
+          new JsonEventSerializer(jsonMapper),
+          EVENTS_PER_SECOND,
+          CYCLE_PADDING_MS
+      );
+      long numWritten = streamGenerator.run(
+          generatedTestConfig.getStreamName(),
+          streamEventWriter,
+          secondsToGenerateFirstRound,
+          FIRST_EVENT_TIME
+      );
+
+      // Trigger early handoff
+      indexer.handoffTaskGroupEarly(
+          generatedTestConfig.getFullDatasourceName(),
+          jsonMapper.writeValueAsString(ImmutableList.of(0, 1))
+      );
+
+      // Load the rest of the data
+      numWritten += streamGenerator.run(
+          generatedTestConfig.getStreamName(),
+          streamEventWriter,
+          secondsToGenerateRemaining,
+          FIRST_EVENT_TIME.plusSeconds(secondsToGenerateFirstRound)
+      );
+
+      // Make sure we consume the rest of the data
```

Comment:
Can we validate that there are a total of 3 tasks returned by the tasks api for the test datasource with one that completed in less than 120s, while the other 2 may be running or have a task completion time of > 120s.

This will prove that the tasks handed off early, which is what this feature is trying to provide. We could increase the supervisors taskDuration in the spec to be high enough to make this test not flaky if needed.